File: python/src/scipp/_bins.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
# @author Simon Heybrock
from typing import Dict, Optional, Sequence, Union

from ._scipp import core as _cpp
from ._cpp_wrapper_util import call_func as _call_cpp_func
from .typing import VariableLike, MetaDataMap


class lookup:
    def __init__(self, func: _cpp.DataArray, dim: str):
        self.func = func
        self.dim = dim


class Bins:
    """
    Proxy for access to bin contents and operations on bins of a variable.

    This class is returned from the `bins` property of variables and should
    generally not be created directly.
    """
    def __init__(self, obj):
        self._obj = obj

    def _data(self):
        try:
            return self._obj.data
        except AttributeError:
            return self._obj

    def __mul__(self, lut: lookup):
        copy = self._obj.copy()
        _cpp.buckets.scale(copy, lut.func, lut.dim)
        return copy

    def __truediv__(self, lut: lookup):
        copy = self._obj.copy()
        _cpp.buckets.scale(copy, _cpp.reciprocal(lut.func), lut.dim)
        return copy

    def __imul__(self, lut: lookup):
        _cpp.buckets.scale(self._obj, lut.func, lut.dim)
        return self

    def __itruediv__(self, lut: lookup):
        _cpp.buckets.scale(self._obj, _cpp.reciprocal(lut.func), lut.dim)
        return self

    @property
    def coords(self) -> MetaDataMap:
        """Coords of the bins"""
        return _cpp._bins_view(self._data()).coords

    @property
    def meta(self) -> MetaDataMap:
        """Coords and attrs of the bins"""
        return _cpp._bins_view(self._data()).meta

    @property
    def attrs(self) -> MetaDataMap:
        """Coords of the bins"""
        return _cpp._bins_view(self._data()).attrs

    @property
    def masks(self) -> MetaDataMap:
        """Masks of the bins"""
        return _cpp._bins_view(self._data()).masks

    @property
    def data(self) -> _cpp.Variable:
        """Data of the bins"""
        return _cpp._bins_view(self._data()).data

    @data.setter
    def data(self, data: _cpp.Variable):
        """Set data of the bins"""
        _cpp._bins_view(self._data()).data = data

    @property
    def constituents(self) -> Dict[str, Union[str, _cpp.Variable, _cpp.DataArray]]:
        """Constituents of binned data, as supported by :py:func:`sc.bins`."""
        begin_end = _cpp.bins_begin_end(self._data())
        return {
            'begin': begin_end[0],
            'end': begin_end[1],
            'dim': _cpp.bins_dim(self._data()),
            'data': _cpp.bins_data(self._data())
        }

    def sum(self) -> Union[_cpp.Variable, _cpp.DataArray]:
        """Sum of each bin.

        :return: The sum of each of the input bins.
        :seealso: :py:func:`scipp.sum` for summing non-bin data
        """
        return _call_cpp_func(_cpp.buckets.sum, self._obj)

    def size(self) -> Union[_cpp.Variable, _cpp.DataArray]:
        """Number of events or elements in a bin.

        :return: The number of elements in each of the input bins.
        """
        return _call_cpp_func(_cpp.bin_size, self._obj)

    def concatenate(
            self,
            other: Optional[Union[_cpp.Variable, _cpp.DataArray]] = None,
            *,
            dim: Optional[str] = None,
            out: Optional[_cpp.DataArray] = None
    ) -> Union[_cpp.Variable, _cpp.DataArray]:
        """Concatenate bins element-wise by concatenating bin contents along
        their internal bin dimension.

        This can be used as a binary operation, or a reduction operation: The
        bins to concatenate are either obtained element-wise from `self`
        and `other`, or, if `dim` but not `other` is given, from all bins
        along the given dimension.

        :param other: Optional input containing bins.
        :param dim: Optional dimension along which to merge bins. If not given
                    and `other` is `None`, the bins are merged along all
                    dimensions.
        :param out: Optional output buffer.
        :raises: If `other` is not binned data.
        :return: The bins of the two inputs merged.
        """
        if other is not None and dim is not None:
            raise RuntimeError(
                "`concatenate` requires `other` or a `dim`, but not both.")
        if other is not None:
            if out is None:
                return _call_cpp_func(_cpp.buckets.concatenate, self._obj, other)
            else:
                if self._obj is out:
                    _call_cpp_func(_cpp.buckets.append, self._obj, other)
                else:
                    out = _call_cpp_func(_cpp.buckets.concatenate, self._obj, other)
                return out
        if out is not None:
            raise RuntimeError("`out` arg not supported for concatenate along dim")
        if dim is not None:
            return _call_cpp_func(_cpp.buckets.concatenate, self._obj, dim)
        raise RuntimeError("Reduction along all dims not supported yet.")


class GroupbyBins:
    """
    Proxy for operations on bins of a groupby object
    """
    def __init__(self, obj):
        self._obj = obj

    def concatenate(self, dim):
        return self._obj.concatenate(dim)


def _events(obj):
    """
    Returns the data underlying the bins stored in the object, or None if the
    object stores dense data.
    """
    if _cpp.is_bins(obj):
        if isinstance(obj, _cpp.Variable):
            return _cpp.bins_data(obj)
        else:
            return _cpp.bins_data(obj.data)
    else:
        return None


def _bins(obj):
    """
    Returns helper :py:class:`scipp.Bins` allowing bin-wise operations
    to be performed or `None` if not binned data.
    """
    if _cpp.is_bins(obj):
        return Bins(obj)
    else:
        return None


def _set_bins(obj, bins: Bins):
    # Should only be used by __iadd__ and friends
    assert obj is bins._obj


def _groupby_bins(obj):
    return GroupbyBins(obj)


def histogram(x: Union[_cpp.DataArray, _cpp.Dataset], *,
              bins: _cpp.Variable) -> Union[_cpp.DataArray, _cpp.Dataset]:
    """Create dense data by histogramming data along all dimension given by
    edges.

    :return: DataArray / Dataset with values equal to the sum
             of values in each given bin.
    :seealso: :py:func:`scipp.bin` for binning data.
    """
    return _call_cpp_func(_cpp.histogram, x, bins)


def bin(x: _cpp.DataArray,
        *,
        edges: Optional[Sequence[_cpp.Variable]] = None,
        groups: Optional[Sequence[_cpp.Variable]] = None,
        erase: Optional[Sequence[_cpp.Variable]] = None) -> _cpp.DataArray:
    """Create binned data by binning input along all dimensions given by edges.

    This does not histogram the data, each output bin will contain a "list" of
    input values.

    At least one argument of ``edges`` and ``groups`` is required.

    :param x: Input data.
    :param edges: Bin edges, one per dimension to bin in.
    :param groups: Keys to group input by one per dimension to group in.
    :param erase: Dimension labels to remove from output.
    :return: Binned ``x``.
    :seealso: :py:func:`scipp.histogram` for histogramming data,
              :py:func:`scipp.bins` for creating binned data based on
              explicitly given index ranges.
    """
    if erase is None:
        erase = []
    if groups is None:
        groups = []
    if edges is None:
        edges = []
    return _call_cpp_func(_cpp.bin, x, edges, groups, erase)


def bins(*,
         data: VariableLike,
         dim: str,
         begin: Optional[_cpp.Variable] = None,
         end: Optional[_cpp.Variable] = None) -> _cpp.Variable:
    """Create a binned variable from bin indices.

    The elements of the returned variable are "bins", defined as views into
    ``data``. The returned variable keeps and manages a copy of ``data``
    internally.

    The variables ``begin`` and ``end`` must have the same dims and shape and
    ``dtype=sc.dtype.int64``. The output dims and shape are given by ``begin``.
    If only ``begin`` is given, each bucket is a slice containing a non-range
    slice of ``data`` at the given indices. If neither ``begin`` nor ``end``
    are given, the output has ``dims=[dim]`` and contains all non-range slices
    along that dimension.

    :param begin: Optional begin indices of bins, used for slicing ``data``.
                  If not provided each row of ``data`` is mapped to a different
                  bin.
    :param end: Optional end indices of bins, used for slicing ``data``. If not
                provided this is assumed to be ``begin + 1``.
    :param dim: Dimension of ``data`` that will be sliced to obtain data for
                any given bin.
    :param data: A variable, data array, or dataset containing combined data
                 of all bins.
    :return: Variable containing data in bins.
    :seealso: :py:func:`scipp.bin` for creating DataArrays based on
              binning of coord value instead of explicitly given index ranges.
    """
    return _call_cpp_func(_cpp.bins, begin, end, dim, data)

File: python/bins.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
/// @file
/// @author Simon Hezbrock
#include "scipp/dataset/bins.h"
#include "scipp/core/except.h"
#include "scipp/dataset/bin.h"
#include "scipp/dataset/bins_view.h"
#include "scipp/dataset/shape.h"
#include "scipp/variable/arithmetic.h"
#include "scipp/variable/cumulative.h"
#include "scipp/variable/rebin.h"
#include "scipp/variable/shape.h"
#include "scipp/variable/util.h"
#include "scipp/variable/variable.h"
#include "scipp/variable/variable_factory.h"

#include "bind_data_array.h"
#include "pybind11.h"

using namespace scipp;

namespace py = pybind11;

namespace {

template <class T>
auto call_make_bins(const py::object &begin_obj, const py::object &end_obj,
                    const Dim dim, T &&data) {
  Variable indices;
  Dimensions dims;
  if (!begin_obj.is_none()) {
    const auto &begin = begin_obj.cast<Variable>();
    dims = begin.dims();
    if (!end_obj.is_none()) {
      const auto &end = end_obj.cast<Variable>();
      indices = zip(begin, end);
    } else {
      indices = zip(begin, begin);
      const auto indices_ = indices.values<scipp::index_pair>();
      const auto nindex = scipp::size(indices_);
      for (scipp::index i = 0; i < nindex; ++i) {
        if (i < nindex - 1)
          indices_[i].second = indices_[i + 1].first;
        else
          indices_[i].second = data.dims()[dim];
      }
    }
  } else if (end_obj.is_none()) {
    const auto one = scipp::index{1} * units::one;
    const auto ones = broadcast(one, {dim, data.dims()[dim]});
    const auto begin = cumsum(ones, dim, CumSumMode::Exclusive);
    indices = zip(begin, begin + one);
  } else {
    throw std::runtime_error("`end` given but not `begin`");
  }
  return make_bins(std::move(indices), dim, std::forward<T>(data));
}

template <class T> void bind_bins(pybind11::module &m) {
  m.def(
      "bins",
      [](const py::object &begin_obj, const py::object &end_obj, const Dim dim,
         const T &data) {
        return call_make_bins(begin_obj, end_obj, dim, T(data));
      },
      py::arg("begin") = py::none(), py::arg("end") = py::none(),
      py::arg("dim"), py::arg("data")); // do not release GIL since using
                                        // implicit conversions in functor
}

template <class T> void bind_bin_size(pybind11::module &m) {
  m.def(
      "bin_size", [](const T &x) { return dataset::bucket_sizes(x); },
      py::call_guard<py::gil_scoped_release>());
}

template <class T> auto bin_begin_end(const Variable &var) {
  auto &&[indices, dim, buffer] = var.constituents<bucket<T>>();
  static_cast<void>(dim);
  static_cast<void>(buffer);
  return py::cast(unzip(indices));
}

template <class T> auto bin_dim(const Variable &var) {
  auto &&[indices, dim, buffer] = var.constituents<bucket<T>>();
  static_cast<void>(buffer);
  static_cast<void>(indices);
  return py::cast(std::string(dim.name()));
}

template <class T>
void bind_bins_map_view(py::module &m, const std::string &name) {
  py::class_<T> c(m, name.c_str());
  bind_common_mutable_view_operators<T>(c);
}

template <class T> void bind_bins_view(py::module &m) {
  py::class_<decltype(dataset::bins_view<T>(Variable{}))> c(
      m, "_BinsViewDataArray");
  bind_bins_map_view<decltype(dataset::bins_view<T>(Variable{}).meta())>(
      m, "_BinsMeta");
  bind_bins_map_view<decltype(dataset::bins_view<T>(Variable{}).coords())>(
      m, "_BinsCoords");
  bind_bins_map_view<decltype(dataset::bins_view<T>(Variable{}).masks())>(
      m, "_BinsMasks");
  bind_bins_map_view<decltype(dataset::bins_view<T>(Variable{}).attrs())>(
      m, "_BinsAttrs");
  bind_data_array_properties(c);
  m.def("_bins_view", [](Variable &var) { return dataset::bins_view<T>(var); });
}

} // namespace

void init_buckets(py::module &m) {
  bind_bins<Variable>(m);
  bind_bins<DataArray>(m);
  bind_bins<Dataset>(m);

  bind_bin_size<Variable>(m);
  bind_bin_size<DataArray>(m);
  bind_bin_size<Dataset>(m);

  m.def("is_bins", variable::is_bins);
  m.def("is_bins",
        [](const DataArray &array) { return dataset::is_bins(array); });
  m.def("is_bins",
        [](const Dataset &dataset) { return dataset::is_bins(dataset); });

  m.def("bins_begin_end", [](const Variable &var) -> py::object {
    if (var.dtype() == dtype<bucket<Variable>>)
      return bin_begin_end<Variable>(var);
    if (var.dtype() == dtype<bucket<DataArray>>)
      return bin_begin_end<DataArray>(var);
    if (var.dtype() == dtype<bucket<Dataset>>)
      return bin_begin_end<Dataset>(var);
    return py::none();
  });

  m.def("bins_dim", [](const Variable &var) -> py::object {
    if (var.dtype() == dtype<bucket<Variable>>)
      return bin_dim<Variable>(var);
    if (var.dtype() == dtype<bucket<DataArray>>)
      return bin_dim<DataArray>(var);
    if (var.dtype() == dtype<bucket<Dataset>>)
      return bin_dim<Dataset>(var);
    return py::none();
  });

  m.def("bins_data", [](py::object &obj) -> py::object {
    auto &var = obj.cast<Variable &>();
    if (var.dtype() == dtype<bucket<Variable>>)
      return py::cast(obj.cast<Variable &>().bin_buffer<Variable>());
    if (var.dtype() == dtype<bucket<DataArray>>)
      return py::cast(obj.cast<Variable &>().bin_buffer<DataArray>().view());
    if (var.dtype() == dtype<bucket<Dataset>>)
      // TODO Provide mechanism for creating sharing view as for DataArray above
      return py::cast(obj.cast<Variable &>().bin_buffer<Dataset>());
    return py::none();
  });

  auto buckets = m.def_submodule("buckets");
  buckets.def(
      "concatenate",
      [](const Variable &a, const Variable &b) {
        return dataset::buckets::concatenate(a, b);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "concatenate",
      [](const DataArray &a, const DataArray &b) {
        return dataset::buckets::concatenate(a, b);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "concatenate",
      [](const Variable &var, const Dim dim) {
        return dataset::buckets::concatenate(var, dim);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "concatenate",
      [](const DataArray &array, const Dim dim) {
        return dataset::buckets::concatenate(array, dim);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "append",
      [](Variable &a, const Variable &b) {
        return dataset::buckets::append(a, b);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "append",
      [](DataArray &a, const DataArray &b) {
        return dataset::buckets::append(a, b);
      },
      py::call_guard<py::gil_scoped_release>());
  buckets.def("map", dataset::buckets::map,
              py::call_guard<py::gil_scoped_release>());
  buckets.def("scale", dataset::buckets::scale,
              py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "sum", [](const Variable &x) { return dataset::buckets::sum(x); },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "sum", [](const DataArray &x) { return dataset::buckets::sum(x); },
      py::call_guard<py::gil_scoped_release>());
  buckets.def(
      "sum", [](const Dataset &x) { return dataset::buckets::sum(x); },
      py::call_guard<py::gil_scoped_release>());

  m.def(
      "bin",
      [](const DataArray &array, const std::vector<Variable> &edges,
         const std::vector<Variable> &groups, const std::vector<Dim> &erase) {
        return dataset::bin(array, edges, groups, erase);
      },
      py::call_guard<py::gil_scoped_release>());
  m.def("bin_with_coords", [](const Variable &data, const py::dict &coords,
                              const std::vector<Variable> &edges,
                              const std::vector<Variable> &groups) {
    std::map<Dim, Variable> c;
    for (const auto [name, coord] : coords)
      c.emplace(Dim(py::cast<std::string>(name)), py::cast<Variable>(coord));
    py::gil_scoped_release release; // release only *after* using py::cast
    return dataset::bin(data, c, std::map<std::string, Variable>{},
                        std::map<Dim, Variable>{}, edges, groups);
  });

  bind_bins_view<DataArray>(m);
}
